client/SDL3/sdl_clip.cpp

```diff
@@ -0,0 +1,765 @@
+      _system(ClipboardCreate()), _event(CreateEventA(nullptr, TRUE, FALSE, nullptr))
+{
+	WINPR_ASSERT(sdl);
+}
+
+sdlClip::~sdlClip()
+{
+	cliprdr_file_context_free(_file);
+	ClipboardDestroy(_system);
+	CloseHandle(_event);
+}
+
+BOOL sdlClip::init(CliprdrClientContext* clip)
+{
+	WINPR_ASSERT(clip);
+	_ctx = clip;
+	clip->custom = this;
+	_ctx->MonitorReady = sdlClip::MonitorReady;
+	_ctx->ServerCapabilities = sdlClip::RecieveServerCapabilities;
+	_ctx->ServerFormatList = sdlClip::ReceiveServerFormatList;
+	_ctx->ServerFormatListResponse = sdlClip::ReceiveFormatListResponse;
+	_ctx->ServerFormatDataRequest = sdlClip::ReceiveFormatDataRequest;
+	_ctx->ServerFormatDataResponse = sdlClip::ReceiveFormatDataResponse;
+
+	return cliprdr_file_context_init(_file, _ctx);
+}
+
+BOOL sdlClip::uninit(CliprdrClientContext* clip)
+{
+	WINPR_ASSERT(clip);
+	if (!cliprdr_file_context_uninit(_file, _ctx))
+		return FALSE;
+	_ctx = nullptr;
+	clip->custom = nullptr;
+	return TRUE;
+}
+
+bool sdlClip::handle_update()
+{
+	if (!_ctx || !_sync)
+		return true;
+
+	clearServerFormats();
+
+	std::string mime_uri_list = "text/uri-list";
+	std::string mime_html = "text/html";
+
+	std::vector<std::string> mime_bitmap = { "image/bmp", "image/x-bmp", "image/x-MS-bmp",
+		                                     "image/x-win-bitmap" };
+	std::string mime_webp = "image/webp";
+	std::string mime_png = "image/png";
+	std::string mime_jpeg = "image/jpeg";
+	std::string mime_tiff = "image/tiff";
+	std::vector<std::string> mime_images = { mime_webp, mime_png, mime_jpeg, mime_tiff };
+
+	std::vector<std::string> clientFormatNames;
+	std::vector<CLIPRDR_FORMAT> clientFormats;
+	if (SDL_HasClipboardText())
+	{
+		clientFormats.push_back({ CF_TEXT, nullptr });
+		clientFormats.push_back({ CF_OEMTEXT, nullptr });
+		clientFormats.push_back({ CF_UNICODETEXT, nullptr });
+	}
+	if (SDL_HasClipboardData(mime_html.c_str()))
+		clientFormatNames.push_back(type_HtmlFormat);
+
+	for (auto& mime : mime_bitmap)
+	{
+		if (SDL_HasClipboardData(mime.c_str()))
+		{
+			clientFormats.push_back({ CF_DIB, nullptr });
+			clientFormats.push_back({ CF_DIBV5, nullptr });
+
+			for (auto& bmp : mime_bitmap)
+				clientFormatNames.push_back(bmp);
+
+			for (auto& img : mime_images)
+				clientFormatNames.push_back(img);
+
+			break;
+		}
+	}
+
+	for (auto& img : mime_images)
+	{
+		if (SDL_HasClipboardData(img.c_str()))
+			clientFormatNames.push_back(img);
+	}
+
+	for (auto& name : clientFormatNames)
+	{
+		clientFormats.push_back({ ClipboardRegisterFormat(_system, name.c_str()), name.data() });
+	}
+
+	std::sort(clientFormats.begin(), clientFormats.end(),
+	          [](const auto& a, const auto& b) { return a < b; });
+	auto u = std::unique(clientFormats.begin(), clientFormats.end());
+	clientFormats.erase(u, clientFormats.end());
+
+	const CLIPRDR_FORMAT_LIST formatList = {
+		.common = { .msgType = CB_FORMAT_LIST, .msgFlags = CB_RESPONSE_OK },
```

Comment:
`CB_RESPONSE_OK` is invalid here (it is only allowed in `*Response` PDUs). In my opinion, the clipboard PDU serialization handling (in `channels/cliprdr/client/cliprdr_main.c`) should fail sending any non-response PDUs, when the `CB_RESPONSE_OK` or `CB_RESPONSE_FAIL` flags are set (because that is not protocol-conform).
(Or have an assertion there that those flags are not used there (asserting correct API usage).)

Reply:
not used in `cliprdr_packet_format_list_new` so no protocol violation.
can print a warning though, that might help spot issues in code.

Reply:
ok, added a warning with 21456dab097d003b2236e164d73fe97c7e1f3923